src/commands/mod.rs

```diff
@@ -582,18 +582,28 @@ struct ResolveArgs {
 
 /// Restore paths from another revision
 ///
+/// With the `-r` option, which is the default, restores a revision from its
+/// parents.
```

Comment:
(Please ignore if this has already been discussed.)

I feel `restore -rREV` is confusing. Unlike `diffedit -rREV` which edits the `REV`, `restore -rREV` sounds like restoring from the `REV` state.

We could say it somehow restores the `REV`, and if the `REV` is the working-copy commit, I agree it's reasonable to "restore" from the parent. However, if the `REV` isn't a working commit, I wouldn't expect that "restore" clears changes in that commit.

Reply:
This is an interesting point. At the very least, I should change the docs; I'm thinking of saying/emphasizing that `jj restore -r` undoes whatever you can see with `jj diff -r`. I'm not sure if that's enough to solve the problem, though.

I agree that this could be surprising, especially to someone unfamiliar with `jj`'s UI. It's also a little dangerous since a new user may not realize what happened when they execute the command, and might lose work as a result.

To a user familiar with `jj diff -r`, `jj rebase -r`, `jj diffedit -r`, I think it should be quite natural that `jj restore -r` acts on the commit you give to it. I added this when I realized that if `jj restore` with no arguments modifies `@`,  I 100% expect to be able to modify any revision in the same way without the need to first `jj edit`. I think that generally, jj should allow this.

There's always the option to rename the option or have a config setting to enable this behavior. There's no rush, I'll think about what's best, perhaps you or others will have more thoughts as well.

Reply:
I think "restore" is named in the traditional HEAD commit + working copy model, just like "checkout". And I feel the proposed behavior is something like `jj resotre --at-rev REV`. (I'm not suggesting we should add `--at-rev` option. I just mean if the verb is "restore", "--at-rev" would be less confusing.)

Implementation-wise, I agree `-r` should modify the specified revision.

Reply:
Should we rename it to `reset`, `empty`, or `clear`? Or too confusing for git users?

Reply:
`jj empty --to` and `jj clear --to` don't sound great. Admittedly, `jj restore --to` and every option I can think of is also a little iffy in this respect, but these two seem extra-bad.

I'm not sure about `reset`. Another option is `refresh`. A weird option is `renew`. 

I'm also not yet sure that renaming the command is the best approach. I like "restore'. We're restoring a revision to its original empty state.

Reply:
We could have `jj replace --to/--from` in place of of `jj restore --to/--from`. (Alternative name: `overwrite`) We could keep `jj restore -r` for restoring a revision (or a portion of it) to an empty revision state. Or, we could use `reset` or one of the other synonyms for that.

Potentially, we could also have an option that makes replace/restore aliases  of each other with the current behavior.

Reply:
I like `restore`, too. `empty`/`clear` doesn't make sense unless user is thinking in a diff mindset. `reset` might work in snapshot model, but yeah, it would be confusing for git users.

`replace`/`overwrite` also work with `--from`/`--to`, but with `-r`, I think they have the same problem as `restore -r`.

Reply:
Exactly, so one possible idea is to split it into two commands, e.g. `replace`  for `--form/--to` and `reset` for `-r` or `overwrite` and `restore`. I don't think `restore -r` is confusing if `restore --to/--from ` does not exist.

OTOH, this is not ideal; if we can find a way to have it make sense as one command, I'd prefer that. Conceptually, here is a way of thinking of what this one command does: it makes the corresponding `jj diff` command show less.

Reply:
I came up with an option I like.

We keep `restore --to/--from` as-is, and simply rename the `-r` option I'm introducing here to `--reset`. As an option, I don't think this is confusing to Git users. Neither is it easy to confuse `--reset` with `--from`. There is no short option for `--reset`.

I see two minor downsides:

1. I'd prefer a word that doesn't start with `r` so that we could have a short option as well. I can't for the life of me think of one. `--clear` is the closest, but I don't find it very clear.
2. Some people might miss `-r/--revision` by analogy with `diff`, `diffedit`, etc. Is there are any complaints, we can have these print an error message. If there are still complaints (which I doubt, as long as we document things properly), we can create an option to make them work the same as `--reset`.

Reply:
I agree with Yuya that it's easy to think that `jj restore -r REV` restores from `REV`. That's unfortunate, because I like the consistency of making it restore from `REV`'s parents.

I quite like `jj reset`, but I think the existence and prevalence of `git reset` unfortunately makes that not a good choice.

I'm not sure that `jj restore --reset REV` is very clear either because there's both "restore" and "reset" in there and it's not clear how the two verbs relate.

I wondered for a while if maybe the use case is simply rare enough that we don't even need this flag, but it does seem a little bad if the user can't reset files in a merge commit without first editing the commit. Hmm, I do think "reset" sounds quite good as a description of what it does.

I'm still not sure what I think we should do about this flag, but I wonder if we could make ourselves more free to choose good names of flags and commands by providing hints to new users when the flag/command is different from other VCSs. For example, we could make `jj restore -r REV` do what Ilya initially proposed while also printing a hint saying something like "Hint: Unlike `hg revert`, this command restores from the parents of the specified revision. Use `jj undo` to undo this operation". If we instead went with `jj reset`, we could say something like "Hint: Unlike `git reset`, this command resets file contents in the specified commit; it does not check out the commit. Use `jj undo` to undo this operation". We would need to make it possible to turn those hints off, on a per-hint bases. Would that be too annoying for the user?

Reply:
Yes, `jj restore --reset REV` is a little awkward. I think I'm OK with that because it's not going to be the most commonly used operation, and `reset` is a good word for it. I like `reset` as a command name less because `jj reset --to` sounds a little weird.

The thing that makes `jj restore -r` particularly dangerous, in my mind, is that it's easy to not notice that you messed up another commit until much later. At that point, even figuring out which command to undo might not be easy or might cause conflicts. That's why I suggested that, if we do want to use `-r`, we should disable it by default. Forcing a user to type `--reset` also seemed like a good compromise to me.

Regarding hints, I think it very much depends on what the command is and how likely we are to correctly detect that the user is confused. I would prefer doing that for specific flag combinations rather than entire commands, unless we expect the command to be rarely used for beginner-to-intermediate users.

Reply:
> I don't like reset as a command name as much because jj reset --to sounds a little weird.

fwiw, I feel `restore --to` is also a bit confusing (because I could read it as "restore the current revision to the given state.) `restore --into` is less confusing, but it doesn't work for immutable `diff` command.

Ignoring git and command consistency, I like `reset`/`-r`/`--from`/`-r --from` (without `--to`) combination. If we want to add an option to `jj restore`, I slightly prefer `restore --at-rev` (copied from `hg rename`.)

Reply:
From the viewpoint of not knowing any of the commands, it is definitely confusing to have so many options. I think the problem that Git has is that it is not obvious which commands affect the index and which affect the working copy. You just have to memorize. But `jj` doesn't have an index, so that particular problem is overcome. However, it should be obvious to the user what each command does. If there are some that are so similar, then it should just be one command with flags.
I really dislike having to refresh my memory of the difference between reset, revert, restore, rebase, checkout, etc. for all the different systems I use.
My personal opinion is that the goal of `jj` is to make it easier on the user, so it should have a lot fewer commands. Take the cognitive load off of the user as much as possible.

Reply:
> If we want to add an option to `jj restore`, I slightly prefer `restore --at-rev` (copied from `hg rename`.)

I think I agree with that.

> My personal opinion is that the goal of `jj` is to make it easier on the user, so it should have a lot fewer commands. Take the cognitive load off of the user as much as possible.

Yes, I think we're all trying to do that. I think the basic functionality of `jj restore` is already simpler than git's and similar to hg's, while still being more powerful than both (being able to restore into an arbitrary commit). What we're talking about in this PR is another detail in the feature that git and hg don't have: being able to restore a file in an arbitrary merge commit from the parents. I don't think it's bad to have that functionality, as long as it doesn't get in the way of normal usage, which I don't think it would if we add a new flag.

I think we should try to simplify the interface by keeping commands consistent and orthogonal. Using `-r` for this command would be consistent with `jj diff`. A way to make commands more orthogonal might be to instead remove the `-r` flag and have `jj diff --from X- --to X` work even if X is merge commit, presumably by resolving the commits in `X-` to an auto-merged tree. However, I think `jj diff -r` is so frequently used that it makes sense to have `-r` even if we made `jj diff --from X-` work on merge commits.

Reply:
> while still being more powerful than both (being able to restore into an arbitrary commit)

`jj copy` would probably be a better name since you can "restore" into an earlier commit or in an unrelated commit.

Reply:
`jj copy` might work, but only after we rename `jj move` to something else. Otherwise it sounds like it would copy the diff from one change to another.

By the way, `hg diff` doesn't have a `-r REV` flag but has a `-c REV` to show the diff caused by a revision. We could copy that "pattern" (not really a pattern, because it's only used for `hg diff`, I think). I think that would mean that we rename the `-r` argument to `-c` for `move`, `squash`, `unsquash`, `diff`, `diffedit`, `rebase`, and we would keep `-r` for `files` and `cat`. That seems consistent, but I suspect it would not be obvious to the user which flag to use, meaning that they'd be annoyed that they have to remember different flag names for no apparent reason. I thought I'd still mention the idea in case others feel differently.

Reply:
I think it would make sense to use the option name to clarify in the UI whether the command treats a revision as a snapshot of the worktree or as a diff from its parents. So, I like the idea of renaming the `-r` flag in some commands in general.

In terms of particulars, however, I dislike the idea of `-c`, at least as long as it means `--change`. To me, `--change` and `--revision` is too much of a synonym to use both in the UI; we should pick one. Something like `--changes-in` is awkward, but clearer, perhaps we can think of something better.

I also think that `rebase` should be treated specially. `rebase -r` is already only loosely connected to `diffedit -r`. `rebase --changes-in` doesn't make sense to me since `rebase -s` and `-rebase -b` also rebase the changes in revisions.

Regarding `restore`, I'm more confused than ever. I like @yuja's thought about `--into`. I don't like `--at-rev` as much, but it does have some advantages. `restore --changes-in` sounds pretty good to me. At this point, I think my preference would be to abandon `restore` having the same UI as `diff` and use `--into` and `--changes-in` (regardless of whether we use `--changes-in` in other commands). Trying to use `--to` or `-r` would at least print a good error message, or we could let users activate it with a config option.

Fortunately, as I keep reminding myself, there is almost 0 urgency in resolving this.

Reply:
> A way to make commands more orthogonal might be to instead remove the -r flag and have jj diff --from X- --to X work even if X is merge commit, presumably by resolving the commits in X- to an auto-merged tree.

We talked more about this on Discord today. We're thinking that something like `jj diff --from 'all!X-' --to X` would make sense as a way of making it easy to have a revset resolve to multiple revisions while still not making it the default so it happens accidentally (maybe the user didn't realize that `X-` (or `X` for that matter) resolved to multiple revisions).

Reply:
Even if we create a way to represent the effect of `jj diff -r` with `--from` and `--to`, we might still want to have a short `-r` flag. Writing `jj diff -r X` is very useful, and having to write `jj diff --from 'all!X-' --to X` instead would be a chore. If we remove `-r`, I'm worried that 95% of uses of `all!` would then be of the form `jj diff --from 'all!X-' --to X`.

This would be just as true for `diffedit`. I think this is even true for `resolve`, though it is clearly less vital in this case.

Reply:
Yep, I completely agree.

Reply:
Re option name: `--revision-from-parents` is the best I came up with so far.

Re `all!`, it seems we settled on `all:` in https://github.com/martinvonz/jj/pull/1911.